### PR TITLE
ssl: allow openssl engine support to be optional

### PR DESCRIPTION
**Pull request by marcuscaisey**
OpenSSL can be built without `ENGINE` support. This PR guards all references to the `ENGINE` API using `OPENSSL_NO_ENGINE` so that the openssl adapters can still be used in this case.

To verify these changes, i've done two things:
1. Guarded the engine related openssl adapter unit tests with `OPENSSL_NO_ENGINE` and added another test artefact which is built with `OPENSSL_NO_ENGINE`
2. Built [Azure/azure-uamqp-python](https://github.com/Azure/azure-uamqp-python) with an OpenSSL with no engine support and the modified shared utilities. Ran the samples [found here](https://github.com/Azure/azure-uamqp-python/tree/master/samples) which ran against Azure Event Hubs*. All passed.

### CODE REVIEW

**Review by CIPop — APPROVED** [review]
Thank you @marcuscaisey . I appreciate the new UTs to validate both cases!

**Comment by CIPop** [comment]
```
[ 61%] Building C object tests/x509_openssl_ut/no_engine/CMakeFiles/x509_openssl_ut_no_engine_exe.dir/main.c.o
In file included from /Users/jenkins/VSTS_agent/_work/19/s/tests/x509_openssl_ut/x509_openssl_ut.c:37:
/usr/local/opt/openssl/include/openssl/engine.h:71:4: error: ENGINE is disabled.
#  error ENGINE is disabled.
   ^
In file included from /Users/jenkins/VSTS_agent/_work/19/s/adapters/x509_openssl.c:14:
/usr/local/opt/openssl/include/openssl/engine.h:71:4: error: ENGINE is disabled.
#  error ENGINE is disabled.
   ^
[ 61%] Building C object tests/x509_openssl_ut/engine/CMakeFiles/x509_openssl_ut_engine_exe.dir/__/__/__/src/xlogging.c.o
1 error generated.
make[2]: *** [tests/x509_openssl_ut/no_engine/CMakeFiles/x509_openssl_ut_no_engine_exe.dir/__/__/__/adapters/x509_openssl.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

**Comment by CIPop** [comment]
@marcuscaisey The OSX build failed with this pragma  : /usr/local/opt/openssl/include/openssl/engine.h:71:4: error: ENGINE is disabled.
AFAIK our gate uses OpenSSL 1.0.2 so it's this line: https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/crypto/engine/engine.h#L71

Please let me know if you plan to make a change. (Gate is blocked by this build failure.)

File: tests/x509_openssl_ut/x509_openssl_ut.c [hunk 1]
```diff
@@ -117,10 +117,12 @@ MOCKABLE_FUNCTION(, long, SSL_CTX_ctrl, SSL_CTX*, ctx, int, cmd, long, larg, voi
 MOCKABLE_FUNCTION(, unsigned long, ERR_peek_last_error);
 MOCKABLE_FUNCTION(, void, ERR_clear_error);
 
+#ifndef OPENSSL_NO_ENGINE
```
Comment (CIPop) [inline tests/x509_openssl_ut/x509_openssl_ut.c:120]: Can't add a comment there but I think you should add a `#ifndef` on line  37 (`#include "openssl/engine.h"`) to fix the OSX bug.

Comment (marcuscaisey) [reply tests/x509_openssl_ut/x509_openssl_ut.c:120]: aha! that would make sense, thanks. i'll try and get back to this soon 👍🏻

Comment (marcuscaisey) [reply tests/x509_openssl_ut/x509_openssl_ut.c:122]: I've guarded that include now, hopefully everything is all good now 🙏

**Comment by marcuscaisey** [comment]
oh no 😢 i'll pull the repo down on my mac (hopefully some time this week) and run the tests and fix them. that will be more efficient than pushing and waiting for the CI build to fail 🙏

**Comment by CIPop** [comment]
Thank you @marcuscaisey ! The build that is failing is OSX with OpenSSL. I remember I had to rebuild OpenSSL from sources to make it work.

There are instructions here: https://github.com/Azure/azure-iot-sdk-c/blob/main/doc/devbox_setup.md#set-up-a-macos-mac-os-x-development-environment although there is always the possibility of having these out of date.

Here are the exact errors:

```
In file included from /Users/jenkins/VSTS_agent/_work/19/s/adapters/x509_openssl.c:14:
/usr/local/opt/openssl/include/openssl/engine.h:71:4: error: ENGINE is disabled.
#  error ENGINE is disabled.
   ^
1 error generated.
[ 58%] Building C object tests/connectionstringparser_ut/CMakeFiles/connectionstringparser_ut_exe.dir/__/real_test_files/real_strings.c.o
make[2]: *** [tests/x509_openssl_ut/no_engine/CMakeFiles/x509_openssl_ut_no_engine_exe.dir/__/__/__/adapters/x509_openssl.c.o] Error 1


[...]


Scanning dependencies of target socketio_connect
make[1]: *** [tests/x509_openssl_ut/no_engine/CMakeFiles/x509_openssl_ut_no_engine_exe.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....


[...]


make: *** [all] Error 2
##[error]Bash exited with code '2'.
Finishing: Build

File: adapters/x509_openssl.c [hunk 2]
```diff
@@ -219,6 +219,7 @@ int x509_openssl_add_pem_file_key(SSL_CTX* ssl_ctx, const char* x509privatekey)
     return result;
 }
 
+#ifndef OPENSSL_NO_ENGINE
```
Comment (CIPop) [inline adapters/x509_openssl.c:222]: I believe the error is now caused by this file. 
I would do a search for `#include "openssl/engine.h"` and add the `#ifdef` everywhere to make the older OpenSSL version compile.

**Comment by marcuscaisey** [comment]
amazing thanks @CIPop, hopefully guarding that include in `x509_openssl.c` fixes the build 🙏

**Comment by ewertons** [comment]
This PR was replaced by #628

**Pull request closed without merging**